---
title: "Asynchronous JavaScript: Callback Function and Promise API"
date: "2019-07-03 11:27:00"
author: "Han Sim"
category: "JavaScript | Web | Node.js"
tags:
  - JavaScript
  - Asynchronous
  - Promise
  - Callback
  - Ajax
  - Async-Await
---

> “A promise is an object that may produce a single value some time in the future”

# Asynchronous

In JavaScript, lots of operation can be done in an `asynchronous` way. To understand `asynchronous programming`, I used this resource: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous. **Asynchronous** means JavaScript does _not_ stop executing the code while it is executing a line of code; it just starts executing the next code.

## Examples 1: jQuery `ajax` communication

```JavaScript
function getData() {
	var data;
	$.get('https://api.com/products/1', function (res) {
		data = res;
	});
	return data;
}

console.log(getData());
```

Because JavaScript is asynchronous, `console.log(getData());` line returns `undefined`. It does not wait for `get` communication, instead, it just returns `data` variable which is undefined yet.

## Example 2: `setTimeout()`

This is probably the most common example that I can find anywhere.

```JavaScript
// #1
console.log('Hello');
// #2
setTimeout(function () {
	console.log('Bye');
}, 3000);
// #3
console.log('Hello Again');
```

Because `setTimeout` is asynchronous as well, a browser runs `console.log('Hello Again');` before it runs `setTimeout(...)`.

```
Hello
Hello Again
Bye
```

# Solution: `Callback Function`

A callback is a function that is to be executed after another function has finished executing — hence the name ‘call back’. (https://codeburst.io/javascript-what-the-heck-is-a-callback-aba4da2deced)

With callback function, we can bypass Asynchronous and let it execute the way we want. This is how I solve the `ajax` call problem I mentioned above.

```JavaScript
//create callback function
function doHomework(subject, callback) {
  alert(`Start homework: ${subject}`); //this will be executed first
  callback(); //second
}

//call it
doHomework('math', function() {
  alert("Finished");
}); 
```

## jQuery example

```JavaScript
function getData(callback) {
	$.get('https://api.com/products/1', function (res) {
		callback(res); //pass 'res' from the server to callback function
	});
}

getData(function(data) {
  console.log(data);
})
```

## Another example: `setState()` in **React**

How to setState() correctly? https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

```JavaScript
// Wrong
this.setState({
  counter: this.state.counter + this.props.increment,
});

// Correct
this.setState((state, props) => ({
  counter: state.counter + props.increment
}));

//or.. (without Arrow Syntax)
// Correct
this.setState(function(state, props) {
  return {
    counter: state.counter + props.increment
  };
});
```

## Another Problem: `Callback Hell`

However, if there is a lot of logic that needs to be executed synchronously, using callback function can be really messy as hell. This is called **Callback Hell**.

Ref: https://blog.hellojs.org/asynchronous-javascript-from-callback-hell-to-async-and-await-9b9ceb63c8e8

To solve this problem, we need **Promise** or **async/await**. I'm mainly discussing about **Promise API** today. I'll post about async/await one day.

# Promise API

> The Promise object represents the eventual completion (or failure) of an asynchronous operation, and its resulting value.
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

```JavaScript
var promise1 = new Promise(function(resolve, reject) {
  setTimeout(function() {
    resolve('foo');
  }, 300);
});

promise1.then(function(value) {
  console.log(value);
  // expected output: "foo"
});

console.log(promise1);
// expected output: [object Promise]
```

## Promise is in one of these states:

- pending: initial state, neither fulfilled nor rejected.
- fulfilled: meaning that the operation completed successfully.
- rejected: meaning that the operation failed.

### Pending

If we call `new Promise()` method, it is now in the `Pending` state. **When we call `new Promise()` method, we can access to `resolve` and `reject` as a paramter of Callback Function.**

```JavaScript
new Promise(function (resolve, reject) {
  // ...
});
```

### Fullfilled

```JavaScript
new Promise(function (resolve, reject) {
  resolve();
});
```

When it executes the parameter `resolve` of Callback Function, it is now in the `Fulfilled` state. If the promise is in this state, we can get results by using `then()`.

```JavaScript
function getData() {
  return new Promise(function (resolve, reject) {
    var data = 100;
    resolve(data);
  });
}

//get result by using "then"
getData().then(function (resolvedData) {
  console.log(resolvedData); // 100
});

```

> If the promise is `fulfilled`, we can say it is **complete** now.

### Rejected

When it executes the parameter `reject` of Callback Function, it is now in the `Rejected` state. If the promise is in this state, we can get results by using `catch()`.

```JavaScript
new Promise(function (resolve, reject) {
  reject();
});
```

We can get the reason of the error.

```JavaScript
getData().then().catch(function (err) {
  console.log(err); // Error: Request is failed
});
```

## Example1. Ajax Communication with Promise

```JavaScript
function getData() {
  return new Promise(function (resolve, reject) {
    $.get('api.com/1', function (response) {
      if (response) resolve(response);
      reject(new Error("Request is failed"));
    });
  });
}

getData().then(function (data) {
  console.log(data); // response
}).catch(function (err) {
  console.error(err); // why error?
});
```

## Example2. real code in Node.js

This is from my full-stack web development project: https://blog.hansim.dev/full-stack-web-development-with-express-and-handlebars (**Node.js/Express**)

Briefly, in `data-service.js` file, I have helper functions that handle data using `sequelize`. I used `Promise` as this needs to be synchronized (ajax communication).

This is **a chain of two promises**. If `sequelize.sync()` is `resolved`, `then` we resolve this Promise. If this promise is `resolved` too, the Node.js/Express server finally renders "employees" page using Template Engines.

```JavaScript
module.exports.getAllEmployees = () => {
    return new Promise((resolve, reject) => {
        sequelize.sync()
        .then(() => { resolve(Employee.findAll()); })
        .catch((err) => { reject("No result returned. " + err); });
    });
}
```

I called the `getAllEmployees()` function in `server.js`. If the promise is resolved, it renders "employees" with `data`(response). Otherwise, it catches the error and renders "employees" with error messages.

```JavaScript
app.get('/employees', (req, res) => {
  dataService.getAllEmployees().then((data) => {
      res.render("employees", {employees: data});
  }).catch((err) => {
      res.render("employees", {message: "no results"});
  });
});
```